src/common/nvlist_utils.c

```diff
@@ -436,3 +436,378 @@ PyObject *py_dump_nvlist(nvlist_t *nvl, boolean_t json)
+	PyObject *psroot, *pstype, *pspath;
+
+	if (topology == NULL || !PyList_Check(topology)) {
+		PyErr_SetString(PyExc_TypeError,
+		    "Expected a list for \'topology\'");
+		return (NULL);
+	}
+
+	nvl = fnvlist_alloc();
+	fnvlist_add_string(nvl, "type", "root");
+
+	pskroot = PyUnicode_FromString("root");
+	if (!pskroot) {
+		PyErr_SetString(PyExc_RuntimeError,
+		    "Failed to create Python String");
+		return (NULL);
+	}
+	psktype = PyUnicode_FromString("type");
+	if (!psktype) {
+		Py_DECREF(pskroot);
+		PyErr_SetString(PyExc_RuntimeError,
+		    "Failed to create Python String");
+		return (NULL);
+	}
+	pskdev = PyUnicode_FromString("devices");
+	if (!pskdev) {
+		Py_DECREF(pskroot);
+		Py_DECREF(psktype);
+		PyErr_SetString(PyExc_RuntimeError,
+		    "Failed to create Python String");
+		return (NULL);
+	}
+	size = (uint64_t)PyList_Size(topology);
+
+	for (uint64_t i = 0; i < size; ++i) {
+		item = PyList_GetItem(topology, i);
+		if (!PyDict_Check(item)) {
+			PyErr_SetString(PyExc_TypeError,
+			    "Expected Dictionaries in \'topology\'");
+			goto fail;
+		}
+		psroot = PyDict_GetItem(item, pskroot);
+		if (!PyUnicode_Check(psroot)) {
+			PyErr_SetString(PyExc_TypeError,
+			    "Expected a String for key \'root\'");
+			goto fail;
+		}
+		pstype = PyDict_GetItem(item, psktype);
+		if (!PyUnicode_Check(pstype)) {
+			PyErr_SetString(PyExc_TypeError,
+			    "Expected a String for key \'type\'");
+			goto fail;
+		}
+		devs = PyDict_GetItem(item, pskdev);
+		if (!PyList_Check(devs)) {
+			PyErr_SetString(PyExc_TypeError,
+			    "Expected a List for key \'devices\'");
+			goto fail;
+		}
+		ctroot = PyUnicode_AsUTF8(psroot);
+		if (!ctroot) {
+			PyErr_SetString(PyExc_RuntimeError,
+			    "Failed to create C string from Python string");
+			goto fail;
+		}
+		cttype = PyUnicode_AsUTF8(pstype);
+		if (!cttype) {
+			PyErr_SetString(PyExc_RuntimeError,
+			    "Failed to create C string from Python string");
+			goto fail;
+		}
+
+		strlcpy(croot, ctroot, sizeof(croot));
+		strlcpy(ctype, cttype, sizeof(ctype));
+		to_lower(croot);
+		to_lower(ctype);
+		if (strcmp(croot, "data") == 0 ||
+		    strcmp(croot, "dedup") == 0 ||
+		    strcmp(croot, "log") == 0 ||
+		    strcmp(croot, "special") == 0) {
+			if (strcmp(ctype, "stripe") == 0)
+				tlc += (uint64_t)PyList_Size(devs);
+			else
+				tlc += 1;
+		} else if (strcmp(croot, "spare") == 0) {
+			if (strcmp(ctype, "stripe") != 0) {
+				PyErr_SetString(PyExc_TypeError,
+				    "Spare devices can only be of type stripe");
+				goto fail;	
+			}
+			nsp = PyList_Size(devs);
+		} else if (strcmp(croot, "cache") == 0) {
+			if (strcmp(ctype, "stripe") != 0) {
+				PyErr_SetString(PyExc_TypeError,
+				    "Cache devices can only be of type stripe");
+				goto fail;	
+			}
+			nl2c = PyList_Size(devs);
+		} else {
+			PyErr_SetString(PyExc_TypeError,
+			    "Unknown VDEV type for key \'root\'");
+			goto fail;
+		}
+	}
+
+	if (tlc > 0)
+		child = malloc(tlc * sizeof(nvlist_t *));
```

Comment:
Where practical we should use the python memory interface rather than normal calloc / malloc.

https://docs.python.org/3/c-api/memory.html#memory-interface

Reply:
Unless of course if the nvlist API is taking ownership of the memory.

Reply:
We can use `PyMem_Malloc` here. I have updated accordingly.